articles/051_actions.md

```diff
@@ -0,0 +1,95 @@
+
+To satisfy these use cases, ROS provides a third communication paradigm known as "actions". An action is a goal-oriented request that occurs asynchronously to the requester, is typically (but not necessarily) longer-running than immediate, can be cancelled or replaced during execution, and has a server that provides feedback on execution progress.
+
+This document defines how actions are specified, what they look like to ROS users (both node developers and system integrators)
+
+## Action specification
+
+Actions are specified using a form of the ROS Message IDL. The specification contains three sections, each of which is a message specification:
+
+1. Goal
+
+1. Result
+
+1. Feedback
+
+Any of these sections may be empty.
+
+Between the three sections is a line containing three hyphens, `---`.
+
+Action specifications are stored in a file ending in `.action`. There is one action specification per `.action` file.
+
+An example action specification [taken from the actionlib wiki] is shown below.
+
+```
+# Define the goal
+uint32 dishwasher_id  # Specify which dishwasher we want to use
+---
+# Define the result
+uint32 total_dishes_cleaned
+---
+# Define a feedback message
+float32 percent_complete
+uint32 number_dishes_cleaned
+```
+
+## Serving and using actions
+
+Actions are a first-class citizen in the ROS API, alongside topics and services.
+
+Action clients will use an API that provides a proxy object for the action. This will be a templated class, using the action class generated from the action specification as the template parameter. The client shall create an instance of this class, providing the address of the intended action server. Each instance of this class can only be related to one action server. Methods of the class will provide facilities for sending a goal to the action server, receiving a result, and getting feedback.
+
+Action servers will use an API that provides a templated server class, using the action class generated from the action specification as the template parameter. The node implementer will create a function that implements the action's behaviour, create an instance of the templated server class, and bind the implementing function to the server. The implementing function will receive as one of its parameters the received goal message, and as another parameter the action server instance. The implementation shall use the action server instance to provide progress feedback and to report the result and success/failure/error status of the action's execution.
+
+Actions may be used from or served by real-time nodes. Therefore the actions API must be real-time capable.
```

Comment:
This is a good goal, but we'll have to make some more qualifications on this most likely. We'd also have to make sure services are real-time, which I'm not sure they necessarily are at the moment.

Reply:
Not all use cases would necessarily need to be supported in realtime. If there's an asynchronous API such that the realtime thread can poll/pump the action client/server without blocking that could be made realtime safe. I'd suggest qualifying this assertion to suggest that there's at least one approach that's realtime safe.